File: python/techai_webutils/src/techai_webutils/clients/messaging/memory/subscriber.py
"""In-memory ``MessageConsumer`` for no-infra builds/tests.

``InMemorySubscriber`` reads messages from an ``InMemoryBroker`` instead of SQS, so a worker can
build and unit-test the consume path with no ElasticMQ/SQS. Mirrors Go's ``messaging/memory``.
Owns no external resource; ``close`` stops the receive loop.
"""

from __future__ import annotations

import asyncio
import logging
from typing import TYPE_CHECKING, Self

from techai_webutils.core.interfaces.messaging import MessageConsumer

if TYPE_CHECKING:
    from types import TracebackType

    from techai_webutils.clients.messaging.memory.broker import InMemoryBroker
    from techai_webutils.core.interfaces.messaging import MessageHandler

logger = logging.getLogger(__name__)

_POLL_TIMEOUT = 0.05
"""Seconds each queue read blocks so the loop can observe ``close`` promptly without a sentinel."""


class InMemorySubscriber(MessageConsumer):
    """A ``MessageConsumer`` that reads from a shared ``InMemoryBroker`` (dev/test)."""

    def __init__(self, broker: InMemoryBroker) -> None:
        """Consume from ``broker``; a publisher built from the same broker feeds this subscriber."""
        self._broker = broker
        self._closed = asyncio.Event()

    async def __aenter__(self) -> Self:
        """No resource to acquire; return self."""
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: TracebackType | None,
    ) -> None:
        """Stop the receive loop on context exit."""
        await self.close()

    async def subscribe(self, topic: str, handler: MessageHandler) -> None:
        """Deliver each message on ``topic`` to ``handler`` until ``close`` is called."""
        queue = self._broker.queue(topic)
        while not self._closed.is_set():
            # The poll timeout is the periodic wake that lets the loop observe ``close``; it is a
            # normal control signal, not a message, so it is NOT passed to the handler.
            try:
                message = await asyncio.wait_for(queue.get(), timeout=_POLL_TIMEOUT)
            except TimeoutError:
                continue
            # A handler failure is dropped so one bad message does not kill the loop, mirroring the
            # Go subscriber (which ignores the handler's error return). The in-memory broker has no
            # redelivery, so this is a stub's best effort, not at-least-once.
            try:
                await handler(message)
            except Exception:
                # Best-effort stub: log and keep consuming past a bad message (the in-memory broker
                # has no redelivery), mirroring the Go subscriber's ignore-the-error semantics.
                logger.debug("in-memory subscriber dropped a failing message", exc_info=True)

    async def close(self) -> None:
        """Signal the receive loop to stop after the current message."""
        self._closed.set()

File: go/clients/messaging/memory/subscriber.go
package memory

import (
	"context"
	"sync"

	"github.com/gt-tech-ai/knowledge-engine/go/core/interfaces"
)

// Compile-time interface assertion.
var _ interfaces.MessageConsumer = (*Subscriber)(nil)

// Subscriber consumes messages from a Broker instead of SQS (dev/test; no external broker).
type Subscriber struct {
	// broker is the in-process broker messages are consumed from.
	broker *Broker
	// done is closed by Close to stop the receive loop.
	done chan struct{}
	// once ensures done is closed at most once, making Close idempotent.
	once sync.Once
}

// NewSubscriber consumes from broker; a Publisher built from the same broker feeds this subscriber.
func NewSubscriber(broker *Broker) *Subscriber {
	return &Subscriber{broker: broker, done: make(chan struct{})}
}

// Subscribe delivers each message on topic to handler until Close is called or ctx is cancelled.
// A handler error is dropped: the in-memory broker has no redelivery, so this is a stub's best
// effort — good enough to exercise the consume path with no SQS, not to model at-least-once.
func (s *Subscriber) Subscribe(
	ctx context.Context,
	topic string,
	handler interfaces.MessageHandler,
) error {
	ch := s.broker.topic(topic)
	for {
		select {
		case msg := <-ch:
			_ = handler(ctx, msg)
		case <-ctx.Done():
			return ctx.Err()
		case <-s.done:
			return nil
		}
	}
}

// Close stops the receive loop (idempotent).
func (s *Subscriber) Close() error {
	s.once.Do(func() { close(s.done) })
	return nil
}
